### PR TITLE
Add `-Zlint-llvm-ir`

### PR DESCRIPTION
**Pull request by dianqk**
This flag is similar to `-Zverify-llvm-ir` and allows us to lint the generated IR.

r? compiler

### CODE REVIEW

**Comment by DianQK** [comment]
I can now use `./x build --stage 2` to find a (maybe) problem:

```
Unusual: noalias argument aliases another argument
  invoke void %6(ptr noundef align 1 %0, ptr noalias noundef readonly align 8 dereferenceable(16) %0)
          to label %15 unwind label %10, !dbg !15
rustc-LLVM ERROR: Linter found errors, aborting. (enabled by --lint-abort-on-error)
```

**Comment by matthiaskrgr** [comment]
:thinking:  I though this can already be done, stabley, with `-Cpasses=lint` ?

**Comment by DianQK** [comment]
> 🤔 I though this can already be done, stabley, with `-Cpasses=lint` ?

I want to run lint the LLVM IR that rustc generates. Of course, I can use it together with `-Cno-prepopulate-passes`.

File: compiler/rustc_session/src/session.rs [hunk 1]
```diff
@@ -652,6 +652,10 @@ impl Session {
         self.opts.unstable_opts.verify_llvm_ir || option_env!("RUSTC_VERIFY_LLVM_IR").is_some()
     }
 
+    pub fn lint_llvm_ir(&self) -> bool {
+        self.opts.unstable_opts.lint_llvm_ir || option_env!("RUSTC_LINT_LLVM_IR").is_some()
```
Comment (bjorn3) [inline compiler/rustc_session/src/session.rs:656]: This env var isn't set anywhere.

**Comment by fmease** [comment]
Sorry for the delay. My understanding is limited unfortunately, I can't tell the difference between `-Cpasses=lint` and `-Zlint-llvm-ir` (apart from the `--lint-abort-on-error`)
r? @nikic

**Comment by DianQK** [comment]
> Sorry for the delay. My understanding is limited unfortunately, I can't tell the difference between `-Cpasses=lint` and `-Zlint-llvm-ir` (apart from the `--lint-abort-on-error`)

The main difference is that with the `-Zlint-llvm-ir` flag, the lint will be executed before optimization, allowing us to lint the IR generated by `rustc`, similar to using the `-Cverify-llvm-ir` flag.

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 2]
```diff
@@ -90,6 +91,9 @@ unsafe fn configure_llvm(sess: &Session) {
         if sess.opts.unstable_opts.print_llvm_passes {
             add("-debug-pass=Structure", false);
         }
+        if llvm_version >= (19, 0, 0) && sess.opts.unstable_opts.lint_llvm_ir {
+            add("-lint-abort-on-error", false);
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:95]: I don't think we should hardcode this option. You can add it via `-Cllvm-args=-lint-abort-on-error` if desired.

**Review by nikic — COMMENTED** [review]
I'm fine with adding the flag. It may be more convenient to use in a larger compilation where `-C no-prepopulate-passes` may be undesirable.

**Review by nikic — APPROVED** [review]

**Pull request merged**